File: backend/src/main/java/com/blog/app/service/PostService.java
package com.blog.app.service;

import java.util.List;

import com.blog.app.exception.BadRequestEception;
import com.blog.app.exception.PostNotFoundException;
import com.blog.app.exception.UserNotExistException;
import com.blog.app.model.Post;

public interface PostService {
	public Post addPost(String email,Post post) throws UserNotExistException;
	public Post updatePost(String email,Post post) throws UserNotExistException ,PostNotFoundException,BadRequestEception;
	public void deletePost( String email,String postId) throws UserNotExistException ,PostNotFoundException,BadRequestEception;

    List<Post> getAllPosts();

    Post getPostById(String id);
    
    public List<Post> getPostsByTopics(List<String> topicNames);
    
}

File: frontend/src/app/services/post/post.service.ts
import { Injectable } from '@angular/core';
import { environment } from '../../../environments/environment';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Post } from 'src/app/model/Post';
@Injectable({
  providedIn: 'root'
})
export class PostService {

  constructor(private http: HttpClient) { }
  addPost(post: any): Observable<Post> {
    return this.http.post<Post>(`${environment.apiUrl}/api/v1/post`, post);
  }
  getAllPost(): Observable<Post[]> {
    return this.http.get<Post[]>(`${environment.apiUrl}/api/v1/post`);
  }
  getPostById(postId: string): Observable<Post> {
    return this.http.get<Post>(`${environment.apiUrl}/api/v1/post/${postId}`);
  }
  updatePost(post: Post) {
    return this.http.put<Post>(`${environment.apiUrl}/api/v1/post`, post);
  }
  deletePost(postId: string) {
    return this.http.delete<Post>(`${environment.apiUrl}/api/v1/post/${postId}`);
  }
}
